### PR TITLE
Workaround golang linux/arm64 link error

### PR DESCRIPTION
**Pull request by yselkowitz**
There is a bug with the golang internal linker for the linux/arm64 target
that has yet to be fixed:

https://github.com/golang/go/issues/40492

### CODE REVIEW

**Review by jstuever — APPROVED** [review]

File: hack/build.sh [hunk 1]
```diff
@@ -42,5 +42,11 @@ if (echo "${TAGS}" | grep -q 'libvirt')
 then
 	export CGO_ENABLED=1
 fi
+if test "$(go env GOARCH)" = "arm64"
```
Comment (staebler) [inline hack/build.sh:45]: Is this change needed when compiling **on** arm64 or **targeting** arm64?

File: hack/build.sh [hunk 2]
```diff
@@ -42,5 +42,11 @@ if (echo "${TAGS}" | grep -q 'libvirt')
...
+then
+	# https://github.com/golang/go/issues/40492
+	LDFLAGS="${LDFLAGS} -linkmode external"
+	export CGO_ENABLED=1
```
Comment (staebler) [inline hack/build.sh:49]: Why does CGO need to be enabled? I am also a bit confused as to why linkmode comes into play at all since the installer only use CGO when enabling libvirt support.

File: hack/build.sh (hunk 1 above)
Comment (yselkowitz) [reply hack/build.sh:45]: Targeting arm64.

File: hack/build.sh (hunk 2 above)
Comment (yselkowitz) [reply hack/build.sh:49]: As described in the golang bug mentioned in the commit message, the golang internal linker does not know how to link a binary as large as the installer for the linux/arm64 target.  Forcing use of the external linker is the only known workaround at the moment until that gets fixed.

**Comment by staebler** [comment]
1. Are you able to do the asset generation when targeting arm64? When I run it, I get the following error.
```
+ go generate ./data
fork/exec /tmp/go-build935493195/b001/exe/assets_generate: exec format error
data/assets.go:2: running "go": exit status 1
```

2. When I build on my dev environment targeting arm64, I get the following errors.
```
+ CGO_ENABLED=1
+ go build -mod=vendor -ldflags ' -X github.com/openshift/installer/pkg/version.Raw=unreleased-master-3751-g801ece65139d7314fd34144e6b1cefc281f833e9 -X github.com/openshift/installer/pkg/version.Commit=801ece65139d7314fd34144e6b1cefc281f833e9 -s -w -linkmode external' -tags ' release' -o bin/openshift-install ./cmd/openshift-install
# runtime/cgo
gcc_arm64.S: Assembler messages:
gcc_arm64.S:28: Error: no such instruction: `stp x29,x30,[sp,'
gcc_arm64.S:32: Error: too many memory references for `mov'
gcc_arm64.S:34: Error: no such instruction: `stp x19,x20,[sp,'
gcc_arm64.S:37: Error: no such instruction: `stp x21,x22,[sp,'
gcc_arm64.S:40: Error: no such instruction: `stp x23,x24,[sp,'
gcc_arm64.S:43: Error: no such instruction: `stp x25,x26,[sp,'
gcc_arm64.S:46: Error: no such instruction: `stp x27,x28,[sp,'
gcc_arm64.S:50: Error: too many memory references for `mov'
gcc_arm64.S:51: Error: too many memory references for `mov'
gcc_arm64.S:52: Error: too many memory references for `mov'
gcc_arm64.S:54: Error: no such instruction: `blr x20'
gcc_arm64.S:55: Error: no such instruction: `blr x19'
gcc_arm64.S:57: Error: no such instruction: `ldp x27,x28,[sp,'
gcc_arm64.S:60: Error: no such instruction: `ldp x25,x26,[sp,'
gcc_arm64.S:63: Error: no such instruction: `ldp x23,x24,[sp,'
gcc_arm64.S:66: Error: no such instruction: `ldp x21,x22,[sp,'
gcc_arm64.S:69: Error: no such instruction: `ldp x19,x20,[sp,'
gcc_arm64.S:72: Error: no such instruction: `ldp x29,x30,[sp],'
```

**Comment by yselkowitz** [comment]
1. go generate never works when cross-compiling, so that is to be expected.
2. I obviously haven't seen such an error in my testing.  With which version are you getting that?

**Comment by yselkowitz** [comment]
ping?

**Comment by staebler** [comment]
>     2. I obviously haven't seen such an error in my testing.  With which version are you getting that?

Version of what? I am using go1.15.2.

```
$ SKIP_GENERATION=y GOARCH=arm64 ./hack/build.sh
+ minimum_go_version=1.14
++ go version
++ cut -d ' ' -f 3
+ current_go_version=go1.15.2
++ version 1.15.2
++ IFS=.
++ printf '%03d%03d%03d\n' 1 15 2
++ unset IFS
++ version 1.14
++ IFS=.
++ printf '%03d%03d%03d\n' 1 14
++ unset IFS
+ '[' 001015002 -lt 001014000 ']'
+ MODE=release
++ git rev-parse --verify 'HEAD^{commit}'
+ GIT_COMMIT=801ece65139d7314fd34144e6b1cefc281f833e9
++ git describe --always --abbrev=40 --dirty
+ GIT_TAG=unreleased-master-3751-g801ece65139d7314fd34144e6b1cefc281f833e9
+ GOFLAGS=-mod=vendor
+ LDFLAGS=' -X github.com/openshift/installer/pkg/version.Raw=unreleased-master-3751-g801ece65139d7314fd34144e6b1cefc281f833e9 -X github.com/openshift/installer/pkg/version.Commit=801ece65139d7314fd34144e6b1cefc281f833e9'
+ TAGS=
+ OUTPUT=bin/openshift-install
+ export CGO_ENABLED=0
+ CGO_ENABLED=0
+ case "${MODE}" in
+ LDFLAGS=' -X github.com/openshift/installer/pkg/version.Raw=unreleased-master-3751-g801ece65139d7314fd34144e6b1cefc281f833e9 -X github.com/openshift/installer/pkg/version.Commit=801ece65139d7314fd34144e6b1cefc281f833e9 -s -w'
+ TAGS=' release'
+ test y '!=' y
+ echo ' release'
+ grep -q libvirt
++ go env GOARCH
+ test arm64 = arm64
+ LDFLAGS=' -X github.com/openshift/installer/pkg/version.Raw=unreleased-master-3751-g801ece65139d7314fd34144e6b1cefc281f833e9 -X github.com/openshift/installer/pkg/version.Commit=801ece65139d7314fd34144e6b1cefc281f833e9 -s -w -linkmode external'
+ export CGO_ENABLED=1
+ CGO_ENABLED=1
+ go build -mod=vendor -ldflags ' -X github.com/openshift/installer/pkg/version.Raw=unreleased-master-3751-g801ece65139d7314fd34144e6b1cefc281f833e9 -X github.com/openshift/installer/pkg/version.Commit=801ece65139d7314fd34144e6b1cefc281f833e9 -s -w -linkmode external' -tags ' release' -o bin/openshift-install ./cmd/openshift-install
# runtime/cgo
gcc_arm64.S: Assembler messages:
gcc_arm64.S:28: Error: no such instruction: `stp x29,x30,[sp,'
gcc_arm64.S:32: Error: too many memory references for `mov'
gcc_arm64.S:34: Error: no such instruction: `stp x19,x20,[sp,'
gcc_arm64.S:37: Error: no such instruction: `stp x21,x22,[sp,'
gcc_arm64.S:40: Error: no such instruction: `stp x23,x24,[sp,'
gcc_arm64.S:43: Error: no such instruction: `stp x25,x26,[sp,'
gcc_arm64.S:46: Error: no such instruction: `stp x27,x28,[sp,'
gcc_arm64.S:50: Error: too many memory references for `mov'
gcc_arm64.S:51: Error: too many memory references for `mov'
gcc_arm64.S:52: Error: too many memory references for `mov'
gcc_arm64.S:54: Error: no such instruction: `blr x20'
gcc_arm64.S:55: Error: no such instruction: `blr x19'
gcc_arm64.S:57: Error: no such instruction: `ldp x27,x28,[sp,'
gcc_arm64.S:60: Error: no such instruction: `ldp x25,x26,[sp,'
gcc_arm64.S:63: Error: no such instruction: `ldp x23,x24,[sp,'
gcc_arm64.S:66: Error: no such instruction: `ldp x21,x22,[sp,'
gcc_arm64.S:69: Error: no such instruction: `ldp x19,x20,[sp,'
gcc_arm64.S:72: Error: no such instruction: `ldp x29,x30,[sp],'
```

**Comment by yselkowitz** [comment]
Ah, you're trying to cross-compile.  Unfortunately, that won't work right now without a linux-aarch64 cross-gcc.  However, the installer is not cross-compiled from one Linux platform to another in OpenShift builds.  This patch allows native builds to succeed, which *is* what is used, and is at least an improvement over being completely unbuildable, until this eventually gets fixed in golang (which won't happen until *at least* 1.16).

**Comment by staebler** [comment]
> Ah, you're trying to cross-compile. Unfortunately, that won't work right now without a linux-aarch64 cross-gcc. However, the installer is not cross-compiled from one Linux platform to another in OpenShift builds. This patch allows native builds to succeed, which _is_ what is used, and is at least an improvement over being completely unbuildable, until this eventually gets fixed in golang (which won't happen until _at least_ 1.16).

Thanks for the explanation.

Also, you do not need to retest. The bot will retest for you since the PR has a lgtm and is approved.
